### PR TITLE
Fix copy errors when using file/blob names with non ascii chars

### PR DESCRIPTION
**Pull request by williexu**
---
Fixes: https://github.com/Azure/azure-cli/issues/4418

-`az storage blob/file url` now returns valid url for non-ascii names
-fixed copy commands
-added tests

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [y ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [y ] Each command and parameter has a meaningful description.
- [y ] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_transformers.py [hunk 1]
```diff
@@ -105,6 +105,9 @@ def transform_url(result):
     """ Ensures the resulting URL string does not contain extra / characters """
     result = re.sub('//', '/', result)
     result = re.sub('/', '//', result, count=1)
-    return result
+    from six.moves.urllib.parse import quote, urlparse, urlunparse  # pylint: disable=import-error
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_transformers.py]: This import can really go at the top of file.
Also, what generates the import error?

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+class StorageCopyTests(StorageScenarioMixin, LiveScenarioTest):
+    @ResourceGroupPreparer()
+    @StorageAccountPreparer()
+    def test_storage_blob_copy_non_ascii(self, storage_account_info):
+        src_container = self.create_container(storage_account_info)
+        dst_container = self.create_container(storage_account_info)
+
+        source_file = self.create_temp_file(16, full_random=False)
+        blobs = ['blobğşŞ', 'blobÉ']
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py]: Do you need to test the (r) sign as well as was reported?

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
+# This Python file uses the following encoding: utf-8
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py]: Can you move this to under the license?

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                                   file_name, src_share).get_output_in_json()
+            self.storage_cmd('storage file copy start -s {} -p "{}" -u {}',
+                             storage_account_info, dst_share, file_name, url).assert_with_checks([
+                                 JMESPathCheck('status', "success")])
+            self.storage_cmd('storage file exists -s {} -p "{}"', storage_account_info,
+                             dst_share, file_name).assert_with_checks(JMESPathCheck('exists', True))
+
+        self.storage_cmd('storage file copy start-batch --destination-share {} --pattern * --source-share {}',
+                         storage_account_info, dst_share, src_share).assert_with_checks(
+                             JMESPathCheck('length(@)', 2))
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py]: Where are the recordings for the tests?

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/blob.py [hunk 5]
```diff
@@ -266,9 +268,18 @@ def _copy_blob_to_blob_container(blob_service, source_blob_service, destination_
...
+        file_url = source_file_service.make_file_url(source_share, source_file_dir or None, source_file_name,
+                                                     sas_token=source_sas)
+    except UnicodeEncodeError:
+        source_file_dir = source_file_dir.encode('utf-8')
+        source_file_name = source_file_name.encode('utf-8')
+        file_url = source_file_service.make_file_url(source_share, source_file_dir or None, source_file_name,
+                                                     sas_token=source_sas)
+    url_parts = urlparse(file_url)
+    quoted_path = quote(url_parts.path, '/()$=\',~')
+    file_url = urlunparse(url_parts[:2] + (quoted_path,) + url_parts[3:])
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/blob.py]: Can we put these 3 lines in a storage util function or something?
It looks like the same thing over and over.
e.g. there is sdkutil.py or storage_url_helpers.py?

Comment (derekbekoe) [reply src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/blob.py]: And also make `'/()$=\',~'` a constant in this util file.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_transformers.py (hunk 1 above)
Comment (williexu) [reply src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_transformers.py]: Not sure, but it seems to be a known problem:
https://github.com/PyCQA/pylint/issues/1640

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py (hunk 4 above)
Comment (williexu) [reply src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py]: I made the tests LiveScenarioTests, so no recordings. 
This was due to me encountering encoding errors when logging messages during playback. 
I didn't want to go about investigating and changing the testbase, hence live-only.
Possible future work.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_transformers.py (hunk 1 above)
Comment (derekbekoe) [reply src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_transformers.py]: okay

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py (hunk 4 above)
Comment (derekbekoe) [reply src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py]: @troydai If you're okay with this only being a live test, I am okay.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py (hunk 3 above)
Comment (williexu) [reply src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py]: Nope, this needs to go above the license in either the first or second line.

Comment (derekbekoe) [reply src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py]: Okay I see from https://www.python.org/dev/peps/pep-0263/

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py [hunk 6]
```diff
@@ -296,11 +296,13 @@ def validate_source_uri(namespace):  # pylint: disable=too-many-statements
...
         query_params.append('snapshot={}'.format(snapshot))
 
+    from six.moves.urllib.parse import quote  # pylint: disable=import-error
+
     uri = 'https://{0}.{1}.{6}/{2}/{3}{4}{5}'.format(
         source_account_name,
         'blob' if valid_blob_source else 'file',
         container if valid_blob_source else share,
-        blob if valid_blob_source else path,
+        quote(blob if valid_blob_source else path, '/()$=\',~'),
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: Why don't you use the util function here?

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_transformers.py [hunk 7]
```diff
@@ -105,6 +105,9 @@ def transform_url(result):
...
+    url_parts = urlparse(result)
+    quoted_path = quote(url_parts.path, '/()$=\',~')
+    return urlunparse(url_parts[:2] + (quoted_path,) + url_parts[3:])
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_transformers.py]: Why don't you use the util function here?

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py (hunk 6 above)
Comment (williexu) [reply src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: I forgot 
...

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py (hunk 4 above)
Comment (troydai) [reply src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py]: Let's make it live only for now.

**Comment by williexu** [comment]
@troydai addressed your concerns

File: src/command_modules/azure-cli-storage/HISTORY.rst [hunk 8]
```diff
@@ -6,6 +6,7 @@ Release History
 2.0.19
 ++++++
 * `storage account create`: defaults --sku to 'Standard_RAGRS'
+* fixed bugs when dealing with file/blob names that include non-ascii chars
```
Comment (troydai) [inline src/command_modules/azure-cli-storage/HISTORY.rst]: Captial, add period.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
+# This Python file uses the following encoding: utf-8
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+import os
+from azure.cli.testsdk import (LiveScenarioTest, ResourceGroupPreparer, StorageAccountPreparer, JMESPathCheck)
```
Comment (troydai) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py]: You don't need the `(` and `)` here.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+class StorageCopyTests(StorageScenarioMixin, LiveScenarioTest):
+    @ResourceGroupPreparer()
+    @StorageAccountPreparer()
+    def test_storage_blob_copy_non_ascii(self, storage_account_info):
+        src_container = self.create_container(storage_account_info)
+        dst_container = self.create_container(storage_account_info)
+
+        source_file = self.create_temp_file(16, full_random=False)
+        blobs = ['blobğşŞ', 'blobÉ®']
```
Comment (troydai) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/tests/test_storage_scenarios_non_ascii_chars.py]: Please open an issue for this shortcoming of VCRpy in https://github.com/azure/azure-python-devtools.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/url_quote_util.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+# --------------------------------------------------------------------------------------------
+
+"""
+Tools to encode url components using python quoting. This is needed to make non-ascii characters safe for requests.
+"""
+
+from six.moves.urllib.parse import urlparse, urlunparse  # pylint: disable=import-error
+from six.moves.urllib.parse import quote as url_quote  # pylint: disable=import-error
+
+SAFE_CHARS = '/()$=\',~'
```
Comment (troydai) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/url_quote_util.py]: Put up the source to this.

**Comment by troydai** [comment]
Will merge when CI passes.

**Pull request merged**